clients/src/main/java/org/apache/kafka/common/record/DefaultRecordBatch.java

```diff
@@ -500,7 +508,19 @@ public Record next() {
                 throw new NoSuchElementException();
 
             readRecords++;
-            return readNext(baseOffset, baseTimestamp, baseSequence, logAppendTime);
+            Record rec = readNext(baseOffset, baseTimestamp, baseSequence, logAppendTime);
+            if (readRecords == numRecords) {
+                // validate the batch
+                boolean isValid = false;
+                try {
+                    readNext(baseOffset, baseTimestamp, baseSequence, logAppendTime);
```

Comment:
This is a little odd. I wonder if we can instead add an abstract `hasRemaining()` method and use `InputStream.available()` and `ByteBuffer.hasRemaining()` to determine whether there are still bytes left. cc @ijuma